### PR TITLE
feat(catalog): allow setting EntityDataParser using CatalogProcessingExtensionPoint

### PR DESCRIPTION
**Pull request by secustor**
## Hey, I just made a Pull Request!

Adds a new function to CatalogProcessingExtensionPoint which allows to provide custom data parser
#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/catalog-backend/src/service/CatalogPlugin.ts [hunk 1]
```diff
@@ -61,6 +64,10 @@ class CatalogProcessingExtensionPointImpl
     this.#placeholderResolvers[key] = resolver;
   }
 
+  setEntityDataParser(parser: CatalogProcessorParser): void {
+    this.entityDataParser = parser;
```
Comment (pjungermann) [inline plugins/catalog-backend/src/service/CatalogPlugin.ts:68]: suggestion:
At a few other places, we throw an exception when such a setter is called twice (e.g., two competing modules). To 
achieve this, the properties are optional and the default is applied when using the value.

Comment (secustor) [reply plugins/catalog-backend/src/service/CatalogPlugin.ts:68]: Make sense to implement the same behaviour, I will look into this.

File: plugins/catalog-backend/src/service/CatalogPlugin.ts [hunk 2]
```diff
@@ -28,18 +28,21 @@ import {
...
 import { PlaceholderResolver } from '../modules';
+import { defaultEntityDataParser } from '../modules/util/parse';
 
 class CatalogProcessingExtensionPointImpl
   implements CatalogProcessingExtensionPoint
 {
   #processors = new Array<CatalogProcessor>();
   #entityProviders = new Array<EntityProvider>();
   #placeholderResolvers: Record<string, PlaceholderResolver> = {};
+  entityDataParser?: CatalogProcessorParser;
```
Comment (sennyeya) [inline plugins/catalog-backend/src/service/CatalogPlugin.ts:45]:
```suggestion
  #entityDataParser?: CatalogProcessorParser;
```
I believe the pattern we're trying to stick with is private internal variables for this extension point, and `#` prefix enforces that.

File: plugins/catalog-backend/src/service/CatalogPlugin.ts [hunk 3]
```diff
@@ -164,6 +176,10 @@ export const catalogPlugin = createBackendPlugin({
         });
         builder.addProcessor(...processingExtensions.processors);
         builder.addEntityProvider(...processingExtensions.entityProviders);
+        builder.setEntityDataParser(
+          processingExtensions.entityDataParser ?? defaultEntityDataParser,
+        );
```
Comment (sennyeya) [inline plugins/catalog-backend/src/service/CatalogPlugin.ts:181]:
```suggestion
        if(processingExtensions.entityDataParser){
            builder.setEntityDataParser(
              processingExtensions.entityDataParser,
            );
        }
```
The [`.build()`](https://github.com/backstage/backstage/blob/patch/v1.22.0/plugins/catalog-backend/src/service/CatalogBuilder.ts#L445) method sets the default down below.

**Comment by benjdlambert** [comment]
ping @backstage/catalog-maintainers

**Comment by secustor** [comment]
Bumping this as it has become stale

**Comment by freben** [comment]
Suggestion: Maybe this would fit best inside a `CatalogModelExtensionPoint` instead of being considered part of the processing per se? https://github.com/backstage/backstage/pull/22568

**Comment by secustor** [comment]
Custom data parser have to emit already known entities, so I'm not thinking this is the right fit. If not the processor endpoint, maybe this should be a reader/parser endpoint.

**Comment by freben** [comment]
The only thing it has to do is to emit something that just vaguely resembles any entity. Basically it needs to be an object, and having `apiVersion`, `kind`, and `metadata.name` - that's all. Those basic requirements are there just to be able to form an entity ref at all, to insert it into the tables that are keyed by that.

**Comment by secustor** [comment]
Good to know! I have thought there is a more strict validation before ingestion in the processor pipelines. 
I feel this is an argument to put this in a `CatalogModelExtensionPoint` but rather which refers to an earlier stage. Maybe `IngestionExtensionPoint` or simply `ExtractionExtensionPoint`?

**Comment by freben** [comment]
I dunno. Ingestion is mostly a thing that's handled by entity providers, which already are on the Processing extension point (whether that's logical or not) :)

I'm still voting `CatalogModelExtensionPoint`.

**Comment by secustor** [comment]
I have changed the used extension point to `CatalogModelExtensionPoint`

**Review by freben — APPROVED** [review]
Seems alright to me. @backstage/catalog-maintainers ?

**Review by alde — APPROVED** [review]
LGTM

**Pull request merged**